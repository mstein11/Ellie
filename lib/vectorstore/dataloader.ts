import {
  MarkdownTextSplitter,
} from 'langchain/text_splitter'
import { Document } from 'langchain/document'

import { randomUUID } from 'crypto'

import legaldata from './data/rulebook-00-legal-srd'
import racesdata from './data/rulebook-01-races-srd'
import classesdata from './data/rulebook-02-classes-srd'
import beyond1stdata from './data/rulebook-03-beyond1st-srd'
import equipmentdata from './data/rulebook-04-equipments-srd'
import featsdata from './data/rulebook-05-feats-srd'
import mechanicsdata from './data/rulebook-06-mechanics-srd'
import combatdata from './data/rulebook-07-combat-srd'
import spellcastingdata from './data/rulebook-08-spellcasting-srd'
import runningdata from './data/rulebook-09-running-srd'
import magicitemsdata from './data/rulebook-10-magic-items-srd'
import monstersdata from './data/rulebook-11-monsters-srd'
import conditionsdata from './data/rulebook-12-conditions-srd'
import godsdata from './data/rulebook-13-gods-srd'
import planesdata from './data/rulebook-14-planes-srd'
import creaturesdata from './data/rulebook-15-creatures-srd'
import npcsdata from './data/rulebook-16-npcs-srd'

type DocumentSlice = {
  data: string
  startInParent: number
  startInDoc: number
  endInDoc: number
  endInParent: number
  length: number
  parentSlice: DocumentSlice | null
  children?: DocumentSlice[]
}

const defaultRegexes = [
  {
    pattern: /.+\n=+\n/g,
    isTable: false
  },
  {
    pattern: /.+\n-+\n/g,
    isTable: false
  },
  {
    pattern: /^#\s.+\n/gm,
    isTable: false
  },
  {
    pattern: /^##\s.+\n/gm,
    isTable: false
  },
  {
    pattern: /^###\s.+\n/gm,
    isTable: false
  },
  {
    pattern: /^####\s.+\n/gm,
    isTable: false
  },
  {
    pattern: /^#####\s.+\n/gm,
    isTable: false
  },
  {
    pattern: /\n\*\*\*.+\*\*\*/g,
    isTable: false
  },
  {
    pattern: /\n\*\*.+\*\*/g,
    isTable: false
  },
  {
    pattern: /(\|.*\|\r?\n\|[-:|]*\|\r?\n)(\|.*\|\r?\n?)*/g,
    isTable: true
  }
]

function getPositionsOfNeedle(inputString: string, needel: RegExp) {
  let positions = []
  let match = needel.exec(inputString)
  while (match) {
    positions.push({ start: match.index, end: match.index + match[0].length })
    match = needel.exec(inputString)
  }
  return positions
}

function handleTable(
  input: DocumentSlice,
  patternConfig: { pattern: RegExp; isTable: boolean }
): DocumentSlice[] {
  const subSlicePositions = getPositionsOfNeedle(
    input.data,
    patternConfig.pattern
  )

  const tableSlices = subSlicePositions.map((subSlicePosition, index) => {
    return {
      data: input.data.substring(subSlicePosition.start, subSlicePosition.end),
      startInParent: subSlicePosition.start,
      startInDoc: input.startInDoc + subSlicePosition.start,
      endInParent: subSlicePosition.end,
      endInDoc: input.startInDoc + subSlicePosition.end,
      length: subSlicePosition.end - subSlicePosition.start,
      parentSlice: input
    }
  })

  const slicesAroundTables = tableSlices
    .map((tableSlice, index) => {
      //beginning until table
      const slices = []
      if (index === 0) {
        if (tableSlice.startInParent !== 0) {
          //first table but not at the beginning
          slices.push({
            data: input.data.substring(0, tableSlice.startInParent),
            startInParent: 0,
            startInDoc: input.startInDoc,
            endInParent: tableSlice.startInParent,
            endInDoc: input.startInDoc + tableSlice.startInParent,
            length: tableSlice.startInParent,
            parentSlice: input
          })
        }
      } else {
        if (tableSlices[index - 1].endInParent !== tableSlice.startInParent) {
          //text between tables
          slices.push({
            data: input.data.substring(
              tableSlices[index - 1].endInParent,
              tableSlice.startInParent
            ),
            startInParent: tableSlices[index - 1].endInParent,
            startInDoc: input.startInDoc + tableSlices[index - 1].endInParent,
            endInParent: tableSlice.startInParent,
            endInDoc: input.startInDoc + tableSlice.startInParent,
            length:
              tableSlice.startInParent - tableSlices[index - 1].endInParent,
            parentSlice: input
          })
        }
      }

      if (index === tableSlices.length - 1) {
        if (tableSlice.endInParent !== input.data.length) {
          //last table but not at the end
          slices.push({
            data: input.data.substring(
              tableSlice.endInParent,
              input.data.length
            ),
            startInParent: tableSlice.endInParent,
            startInDoc: input.startInDoc + tableSlice.endInParent,
            endInParent: input.data.length - 1,
            endInDoc: input.startInDoc + input.data.length - 1,
            length: input.data.length - 1 - tableSlice.endInParent,
            parentSlice: input
          })
        }
      }

      return slices
    })
    .flat()

  tableSlices.push(...slicesAroundTables)
  tableSlices.sort((a, b) => a.startInParent - b.startInParent)

  input.children = tableSlices
  return tableSlices
}

function prepareLvl2Headings(
  input: DocumentSlice,
  patternConfig: { pattern: RegExp; isTable: boolean }
): DocumentSlice[] {
  if (patternConfig.isTable) {
    return handleTable(input, patternConfig)
  }

  const subSlicePositions = getPositionsOfNeedle(
    input.data,
    patternConfig.pattern
  )
  if (subSlicePositions.length === 0) {
    return [input]
  }

  if (
    subSlicePositions.some(subSlicePosition => subSlicePosition.start !== 0)
  ) {
    const startOfFirstSubslice = subSlicePositions.toSorted(
      (a, b) => a.start - b.start
    )[0]
    subSlicePositions.unshift({ start: 0, end: startOfFirstSubslice.start })
  }

  return subSlicePositions.map((subSlicePosition, index) => {
    //this does not work for tables, table can be closed and afterwards there can be text
    let endOfSubSlice = input.data.length
    if (index < subSlicePositions.length - 1) {
      //the start of the next slice is the end of the current slice
      endOfSubSlice = subSlicePositions[index + 1].start - 1
    }

    const subSlice = {
      data: input.data.substring(subSlicePosition.start, endOfSubSlice),
      startInParent: subSlicePosition.start,
      startInDoc: input.startInDoc + subSlicePosition.start,
      endInParent: endOfSubSlice,
      endInDoc: input.startInDoc + endOfSubSlice,
      length: endOfSubSlice - subSlicePosition.start,
      parentSlice: input
    }
    input.children = [subSlice]
    return subSlice
  })
}

export async function loadSource({
  data = [
    racesdata,
    classesdata,
    beyond1stdata,
    equipmentdata,
    featsdata,
    mechanicsdata,
    combatdata,
    spellcastingdata,
    runningdata,
    magicitemsdata,
    monstersdata,
    conditionsdata,
    godsdata,
    planesdata,
    creaturesdata,
    npcsdata,
    legaldata
  ].join('\n'),
  idProvider = () => randomUUID()
} = {}): Promise<{ documents: Document[]; ids: string[] }> {
  const splitter = MarkdownTextSplitter.fromLanguage('markdown', {
    chunkOverlap: 0
  })

  const output = await splitter.createDocuments([data])

  const ids = []
  for (const doc of output) {
    const id = idProvider()
    doc.metadata.id = id
    ids.push(id)
  }
  return { documents: output, ids }
}

export async function loadSourceV2({
  data = [racesdata, classesdata].join('\n'),
  idProvider = () => randomUUID(),
  maxLength = 1000,
  splitterRegexes = defaultRegexes
} = {}): Promise<any> {
  const rootDocSlice: DocumentSlice = {
    data,
    startInParent: 0,
    startInDoc: 0,
    endInDoc: data.length,
    endInParent: data.length,
    length: data.length,
    parentSlice: null,
    children: []
  }

  let currentLevel = [rootDocSlice]
  let counter = 0
  const finalSlices = []

  do {
    if (counter > splitterRegexes.length - 1 || currentLevel.length === 0) {
      finalSlices.push(...currentLevel)
      break
    }

    finalSlices.push(
      ...currentLevel.filter(curLvl => curLvl.data.length <= maxLength)
    )

    currentLevel = currentLevel
      .filter(curlvl => curlvl.data.length > maxLength)
      .map(curLvl => {
        return prepareLvl2Headings(curLvl, splitterRegexes[counter])
      })
      .flat()

    counter++
  } while (currentLevel.some(curlvl => curlvl.data.length > maxLength))

  return finalSlices
}
