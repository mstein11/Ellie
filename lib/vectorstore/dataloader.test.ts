import { loadSource, loadSourceV2, loadSourceV2WithLangchainFormat } from './dataloader'

describe('should test dataloader', () => {
  it('should test dataloader', async () => {
    const result = await loadSource({
      idProvider: () => 'some-test-id-matching-schema'
    })
    expect(result).toMatchSnapshot()
  })

  it('test v2', async () => {
    const result = await loadSourceV2({
      idProvider: () => 'some-test-id-matching-schema'
    })
    const res = result.map((item: any) => {
      return { ...item, parentSlice: null }
    })

    const largeDocs = res
      .filter((item: any) => item.data.length > 1000)
      .sort((a: any, b: any) => a.data.length - b.data.length)
    console.log('large docs: ' + largeDocs.length)
    console.log('all docs: ' + res.length)
    // expect(largeDocs).toMatchSnapshot()

    expect(
      result
        .filter((item: any) => !item.children)
        .sort((a: any, b: any) => a.startInDoc - b.startInDoc)
        .map((item: any) => item.data)
        .join('\n---xxx---\n')
    ).toMatchSnapshot()

    expect(
      res.sort((a: any, b: any) => a.data.length - b.data.length)
    ).toMatchSnapshot()
    expect({
      length: res.length,
      lengthGreater1000: largeDocs.length
    }).toMatchSnapshot()
  })

  it('test splitting for large table', async () => {
    const text = `Fighter
=======

Class Features
--------------

As a fighter, you gain the following class features.

#### Hit Points

**Hit Dice:** 1d10 per fighter level

**Hit Points at 1st Level:** 10 + your Constitution modifier

**Hit Points at Higher Levels:** 1d10 (or 6) + your Constitution modifier per fighter level after 1st

#### Proficiencies

**Armor:** All armor, shields

**Weapons:** Simple weapons, martial weapons

**Tools:** None

**Saving Throws:** Strength, Constitution

**Skills:** Choose two skills from Acrobatics, Animal

Handling, Athletics, History, Insight, Intimidation, Perception, and Survival

#### Equipment

You start with the following equipment, in addition to the equipment granted by your background:
- (*a*) chain mail or (*b*) leather armor, longbow, and 20 arrows
- (*a*) a martial weapon and a shield or (*b*) two martial weapons
- (*a*) a light crossbow and 20 bolts or (*b*) two handaxes
- (*a*) a dungeoneer’s pack or (*b*) an explorer’s pack

#### The Fighter

| Level | Proficiency Bonus | Features                                          |
|-------|-------------------|---------------------------------------------------|
| 1st   | +2                | Fighting Style, Second Wind                       |
| 2nd   | +2                | Action Surge (one use)                            |
| 3rd   | +2                | Martial Archetype                                 |
| 4th   | +2                | Ability Score Improvement                         |
| 5th   | +3                | Extra Attack                                      |
| 6th   | +3                | Ability Score Improvement                         |
| 7th   | +3                | Martial Archetype Feature                         |
| 8th   | +3                | Ability Score Improvement                         |
| 9th   | +4                | Indomitable (one use)                             |
| 10th  | +4                | Martial Archetype Feature                         |
| 11th  | +4                | Extra Attack (2)                                  |
| 12th  | +4                | Ability Score Improvement                         |
| 13th  | +5                | Indomitable (two uses)                            |
| 14th  | +5                | Ability Score Improvement                         |
| 15th  | +5                | Martial Archetype Feature                         |
| 16th  | +5                | Ability Score Improvement                         |
| 19th  | +6                | Ability Score Improvement                         |
| 20th  | +6                | Extra Attack (3)                                  |`

    const result = await loadSourceV2({
      idProvider: () => 'some-test-id-matching-schema',
      data: text
    })

    expect(
      result
        .filter((item: any) => !item.children)
        .sort((a: any, b: any) => a.startInDoc - b.startInDoc)
        .map((item: any) => item.data)
        .join('\n---xxx---\n')
    ).toMatchSnapshot()

    expect(result.length).toBe(7)
  })
})

describe('first level heading', () => {
  it('should not loose text after before heading', async () => {
    const text = `SomeTextBeforeHeading 

The Heading
===========
`

    const result = await loadSourceV2WithLangchainFormat({
      maxLength: 10,
      idProvider: () => 'some-test-id-matching-schema',
      data: text
    })

    expect(result.length).toBe(2)
    expect(result[0].pageContent).toBe('SomeTextBeforeHeading \n\n')
    expect(result[0].metadata.loc.characters.from).toBe(0)
    expect(result[0].metadata.loc.characters.to).toBe(23)
    expect(result[0].metadata.loc.lines.from).toBe(0)
    expect(result[0].metadata.loc.lines.to).toBe(2)
    expect(result[1].pageContent).toBe('The Heading\n===========\n')
    expect(result[1].metadata.loc.characters.from).toBe(24)
    expect(result[1].metadata.loc.characters.to).toBe(47)
    expect(result[1].metadata.loc.lines.from).toBe(2)
    expect(result[1].metadata.loc.lines.to).toBe(4)
  })

  it('should not split if maxLength is long enough', async () => {
    const text = `The Heading #1
    ===========

    Some Text

    The Heading #2
    ==============

    Some other Text`

    const result = await loadSourceV2WithLangchainFormat({
      maxLength: 1000,
      idProvider: () => 'some-test-id-matching-schema',
      data: text
    })

    expect(result.length).toBe(1)
    expect(result[0].pageContent).toBe(text)
    expect(result[0].metadata.loc.characters.from).toBe(0)
    expect(result[0].metadata.loc.characters.to).toBe(104)
    expect(result[0].metadata.loc.lines.from).toBe(0)
    expect(result[0].metadata.loc.lines.to).toBe(8)
  })
})

describe('table splitt logic', () => {
  it('should not loose text after table', async () => {
    const text = `| Level | Proficiency Bonus |
|-------|-------------------|
| 1st   | +2                |
SomeTextAfterTable`

    const result = await loadSourceV2WithLangchainFormat({
      maxLength: 50,
      idProvider: () => 'some-test-id-matching-schema',
      data: text
    })

    expect(result.length).toBe(2)
    expect(result[0].pageContent).toBe(
      '| Level | Proficiency Bonus |\n|-------|-------------------|\n| 1st   | +2                |\n'
    )
    expect(result[0].metadata.loc.characters.from).toBe(0)
    expect(result[0].metadata.loc.characters.to).toBe(89)
    expect(result[0].metadata.loc.lines.from).toBe(0)
    expect(result[0].metadata.loc.lines.to).toBe(3)
    expect(result[1].pageContent).toBe('SomeTextAfterTable')
    expect(result[1].metadata.loc.characters.from).toBe(90)
    expect(result[1].metadata.loc.characters.to).toBe(107)
    expect(result[1].metadata.loc.lines.from).toBe(3)
    expect(result[1].metadata.loc.lines.to).toBe(3)
  })

  it('should not loose text before table', async () => {
    const text = `SomeTextBeforeTable
| Level | Proficiency Bonus |
|-------|-------------------|
| 1st   | +2                |`

    const result = await loadSourceV2WithLangchainFormat({
      maxLength: 50,
      idProvider: () => 'some-test-id-matching-schema',
      data: text
    })

    expect(result.length).toBe(2)
    expect(result[0].pageContent).toBe('SomeTextBeforeTable\n')
    expect(result[0].metadata.loc.characters.from).toBe(0)
    expect(result[0].metadata.loc.characters.to).toBe(19)
    expect(result[0].metadata.loc.lines.from).toBe(0)
    expect(result[0].metadata.loc.lines.to).toBe(1)
    expect(result[1].pageContent).toBe(
      '| Level | Proficiency Bonus |\n|-------|-------------------|\n| 1st   | +2                |'
    )
    expect(result[1].metadata.loc.characters.from).toBe(20)
    expect(result[1].metadata.loc.characters.to).toBe(108)
    expect(result[1].metadata.loc.lines.from).toBe(1)
    expect(result[1].metadata.loc.lines.to).toBe(3)
  })

  it('should not loose text between tables', async () => {
    const text = `| Level | Proficiency Bonus |
|-------|-------------------|
| 1st   | +2                |

SomeTextBetweenTable

| Level | Proficiency Bonus |
|-------|-------------------|
| 2nd   | +2                |`

    const result = await loadSourceV2WithLangchainFormat({
      maxLength: 50,
      idProvider: () => 'some-test-id-matching-schema',
      data: text
    })

    expect(result.length).toBe(3)
    expect(result[0].pageContent).toBe(
      '| Level | Proficiency Bonus |\n|-------|-------------------|\n| 1st   | +2                |\n'
    )
    expect(result[0].metadata.loc.characters.from).toBe(0)
    expect(result[0].metadata.loc.characters.to).toBe(89)
    expect(result[0].metadata.loc.lines.from).toBe(0)
    expect(result[0].metadata.loc.lines.to).toBe(3)
    expect(result[1].pageContent).toBe('\nSomeTextBetweenTable\n\n')
    expect(result[1].metadata.loc.characters.from).toBe(90)
    expect(result[1].metadata.loc.characters.to).toBe(112)
    expect(result[1].metadata.loc.lines.from).toBe(4)
    expect(result[1].metadata.loc.lines.to).toBe(6)
    expect(result[2].pageContent).toBe(
      '| Level | Proficiency Bonus |\n|-------|-------------------|\n| 2nd   | +2                |'
    )
    expect(result[2].metadata.loc.characters.from).toBe(113)
    expect(result[2].metadata.loc.characters.to).toBe(201)
    expect(result[2].metadata.loc.lines.from).toBe(6)
    expect(result[2].metadata.loc.lines.to).toBe(8)
  })
})
