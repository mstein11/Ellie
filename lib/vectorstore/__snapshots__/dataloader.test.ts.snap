// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata should not have doubled ParentHeadings 1`] = `
[
  {
    "metadata": {
      "id": "some-id-random-random-random",
      "loc": {
        "characters": {
          "from": 0,
          "to": 881,
        },
        "lines": {
          "from": 0,
          "to": 38,
        },
      },
      "parentHeadings": [
        "Barbarian
=========
",
        "Class Features
--------------
",
      ],
    },
    "pageContent": "Barbarian
=========

Class Features
--------------

As a barbarian, you gain the following class features.

#### Hit Points

**Hit Dice:** 1d12 per barbarian level

**Hit Points at 1st Level:** 12 + your Constitution modifier

**Hit Points at Higher Levels:** 1d12 (or 7) + your Constitution
modifier per barbarian level after 1st

#### Proficiencies

**Armor:** Light armor, medium armor, shields **Weapons:** Simple
weapons, martial weapons **Tools:** None

**Saving Throws:** Strength, Constitution

**Skills:** Choose two from Animal Handling, Athletics,

Intimidation, Nature, Perception, and Survival

#### Equipment

You start with the following equipment, in addition to the equipment
granted by your background:
- (*a*) a greataxe or (*b*) any martial melee weapon
- (*a*) two handaxes or (*b*) any simple weapon
- An explorer’s pack and four javelins

#### The Barbarian

",
  },
  {
    "metadata": {
      "id": "some-id-random-random-random",
      "loc": {
        "characters": {
          "from": 882,
          "to": 2840,
        },
        "lines": {
          "from": 38,
          "to": 61,
        },
      },
      "parentHeadings": [
        "Barbarian
=========
",
        "Class Features
--------------
",
      ],
    },
    "pageContent": "| Level  | Proficiency Bonus | Features                      | Rages     | Rage Damage |
|--------|-------------------|-------------------------------|-----------|-------------|
| 1st    | +2                | Rage, Unarmored Defense       | 2         | +2          |
| 2nd    | +2                | Reckless Attack, Danger Sense | 2         | +2          |
| 3rd    | +2                | Primal Path                   | 3         | +2          |
| 4th    | +2                | Ability Score Improvement     | 3         | +2          |
| 5th    | +3                | Extra Attack, Fast Movement   | 3         | +2          |
| 6th    | +3                | Path feature                  | 4         | +2          |
| 7th    | +3                | Feral Instinct                | 4         | +2          |
| 8th    | +3                | Ability Score Improvement     | 4         | +2          |
| 9th    | +4                | Brutal Critical (1 die)       | 4         | +3          |
| 10th   | +4                | Path feature                  | 4         | +3          |
| 11th   | +4                | Relentless                    | 4         | +3          |
| 12th   | +4                | Ability Score Improvement     | 5         | +3          |
| 13th   | +5                | Brutal Critical (2 dice)      | 5         | +3          |
| 14th   | +5                | Path feature                  | 5         | +3          |
| 15th   | +5                | Persistent Rage               | 5         | +3          |
| 16th   | +5                | Ability Score Improvement     | 5         | +4          |
| 17th   | +6                | Brutal Critical (3 dice)      | 6         | +4          |
| 18th   | +6                | Indomitable Might             | 6         | +4          |
| 19th   | +6                | Ability Score Improvement     | 6         | +4          |
| 20th   | +6                | Primal Champion               | Unlimited | +4          |

",
  },
  {
    "metadata": {
      "id": "some-id-random-random-random",
      "loc": {
        "characters": {
          "from": 2841,
          "to": 3706,
        },
        "lines": {
          "from": 61,
          "to": 75,
        },
      },
      "parentHeadings": [
        "Barbarian
=========
",
        "Class Features
--------------
",
        "### Rage
",
      ],
    },
    "pageContent": "### Rage

In battle, you fight with primal ferocity. On your turn, you can enter a rage as a bonus action.

While raging, you gain the following benefits if you aren’t wearing heavy armor:

- You have advantage on Strength checks and Strength saving throws.
- When you make a melee weapon attack using Strength, you gain a bonus to the damage roll that increases as you gain levels as a barbarian, as shown in the Rage Damage column of the Barbarian table.
- You have resistance to bludgeoning, piercing, and slashing damage.

If you are able to cast spells, you can’t cast them or concentrate on them while raging.

Your rage lasts for 1 minute. It ends early if you are knocked unconscious or if your turn ends and you haven’t attacked a hostile creature since your last turn or taken damage since then. You can also end your rage on your turn as a bonus action.

",
  },
  {
    "metadata": {
      "id": "some-id-random-random-random",
      "loc": {
        "characters": {
          "from": 3707,
          "to": 3877,
        },
        "lines": {
          "from": 75,
          "to": 77,
        },
      },
      "parentHeadings": [
        "Barbarian
=========
",
        "Class Features
--------------
",
        "### Rage
",
      ],
    },
    "pageContent": "Once you have raged the number of times shown for your barbarian level in the Rages column of the Barbarian table, you must finish a long rest before you can rage again.

",
  },
  {
    "metadata": {
      "id": "some-id-random-random-random",
      "loc": {
        "characters": {
          "from": 3878,
          "to": 4774,
        },
        "lines": {
          "from": 77,
          "to": 90,
        },
      },
      "parentHeadings": [
        "Barbarian
=========
",
        "Class Features
--------------
",
      ],
    },
    "pageContent": "### Unarmored Defense

While you are not wearing any armor, your Armor Class equals 10 + your Dexterity modifier + your Constitution modifier. You can use a shield and still gain this benefit.

### Reckless Attack

Starting at 2nd level, you can throw aside all concern for defense to attack with fierce desperation. When you make your first attack on your turn, you can decide to attack recklessly. Doing so gives you advantage on melee weapon attack rolls using Strength during this turn, but attack rolls against you have advantage until your next turn.

### Danger Sense

At 2nd level, you gain an uncanny sense of when things nearby aren’t as they should be, giving you an edge when you dodge away from danger.

You have advantage on Dexterity saving throws against effects that you can see, such as traps and spells. To gain this benefit, you can’t be blinded, deafened, or incapacitated.
  ",
  },
]
`;
